## Data Cleaning

### Author: Lucia Zou
### Contact: lucia.zouyuebca@gmail.com 
### Date: Oct.18th, 2023

### Table of Contents

- [Introduction](#Introduction)
- [Problem Statement](#Problem-Statement)
- [Methods](#Methods)
- [Feature Overview](#Feature-Overview)
- [Data Import](#Data-Import)
- [Assessment](#Assessment)
- [Handling Missing Values](#Handling-Missing-Values)
- [Extra Cleaning](#Extra-Cleaning)

###

### Introduction

This dataset compiles the spirits purchase data of Iowa Class “E” liquor license holders, detailing products and purchase dates spanning from January 2021 to January 2022. It serves as a valuable resource for analyzing individual product sales at the store level, offering insights into the total spirits sales landscape in Iowa. The objective of this project is to leverage machine learning techniques to forecast future sales for these stores. Our aim is to empower store owners by maximizing their revenue potential and optimizing their operations

###

### Problem Statement

Can we use machine learning models to predict future sales of liquor?

###

### Methods

We utilized the Python programming language along with key libraries like pandas for seamless data manipulation, matplotlib for insightful data visualization, and sklearn for advanced machine learning techniques and data preprocessing. Additionally, we harnessed the power of Jupyter Notebook as our integrated development environment (IDE), enabling interactive programming and dynamic visualization.

###

### Feature Overview

1. invoice_and_item_number: concatenated invoice and line number associated with the liquor order. This provides a unique identifier for the individual liquor products included in the store order.

2. date: date of order.

3. store_number: unique number assigned to the store who ordered the liquor.

4. store_name: name of store who ordered the liquor.

5. address: address of store who ordered the liquor.

6. city: city where the store who ordered the liquor is located.

7. zip_code: zip code where the store who ordered the liquor is located.

8. store_location: location of store who ordered the liquor. The address, city, state and zip code are geocoded to provide geographic coordinates.

9. county_number: iowa county number for the county where store who ordered the liquor is located.

10. county: county where the store who ordered the liquor is located.

11. category: category code associated with the liquor ordered.

12. category_name: category of the liquor ordered.

13. vendor_number: the vendor number of the company for the brand of liquor ordered.

14. vendor_name: the vendor name of the company for the brand of liquor ordered.

15. item_number: item number for the individual liquor product ordered.

16. item_description: description of the individual liquor product ordered.

17. pack: the number of bottles in a case for the liquor ordered.

18. bottle_volume_ml: volume of each liquor bottle ordered in milliliters.

19. state_bottle_cost: the amount that alcoholic beverages division paid for each bottle of liquor ordered.

20. state_bottle_retail: the amount the store paid for each bottle of liquor ordered.

21. bottles_sold: the number of bottles of liquor ordered by the store.

22. sale_dollars: total cost of liquor order (number of bottles multiplied by the state bottle retail).

23. volume_sold_liters: total volume of liquor ordered in liters. 

24. volume_sold_gallons: total volume of liquor ordered in gallons. 

###

### Data Import

In [1]:
import numpy as np
import pandas as pd

import plotly.express as px
from plotly.subplots import make_subplots
import plotly.graph_objs as go
import matplotlib.pyplot as plt
import seaborn as sns

from statsmodels.api import tsa 
import statsmodels.api as sm

In [2]:
df=pd.read_csv('Iowa.csv')
df.head()

invoice_and_item_number        date  store_number  \
0         INV-33179700135  2021-01-04          2576   
1         INV-33196200106  2021-01-04          2649   
2         INV-33184300011  2021-01-04          2539   
3         INV-33184100015  2021-01-04          4024   
4         INV-33174200025  2021-01-04          5385   

                             store_name         address             city  \
0  Hy-Vee Wine and Spirits / Storm Lake  1250 N Lake St       Storm Lake   
1                   Hy-Vee #3 / Dubuque   400 Locust St          Dubuque   
2        Hy-Vee Food Store / Iowa Falls      640 S. Oak       Iowa Falls   
3            Wal-Mart 1546 / Iowa Falls       840 S Oak       Iowa Falls   
4                    Vine Food & Liquor   2704 Vine St.  West Des Moines   

   zip_code                        store_location  county_number      county  \
0   50588.0  POINT (-95.200758 42.65318400000001)           11.0  BUENA VIST   
1   52001.0  POINT (-90.666497 42.49721900000001)           31.0     DUBUQUE   
2   50126.0          POINT (-93.262364 42.508752)           42.0      HARDIN   
3   50126.0          POINT (-93.262446 42.503407)           42.0      HARDIN   
4   50265.0          POINT (-93.741511 41.580206)           77.0        POLK   

   ...  item_number           item_description  pack bottle_volume_ml  \
0  ...        64870  Fireball Cinnamon Whiskey    48              100   
1  ...        65200       Tequila Rose Liqueur    12              750   
2  ...        38008         Smirnoff 80prf PET     6             1750   
3  ...        36648              Caliber Vodka    12              750   
4  ...         4626       Buchanan Deluxe 12YR    12              750   

   state_bottle_cost state_bottle_retail  bottles_sold  sale_dollars  \
0               0.90                1.35            48         64.80   
1              11.50               17.25             4         69.00   
2              14.75               22.13             6        132.78   
3               3.31                4.97            12         59.64   
4              20.99               31.49             2         62.98   

   volume_sold_liters  volume_sold_gallons  
0                 4.8                 1.26  
1                 3.0                 0.79  
2                10.5                 2.77  
3                 9.0                 2.37  
4                 1.5                 0.39  

[5 rows x 24 columns]

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2805307 entries, 0 to 2805306
Data columns (total 24 columns):
 #   Column                   Dtype  
---  ------                   -----  
 0   invoice_and_item_number  object 
 1   date                     object 
 2   store_number             int64  
 3   store_name               object 
 4   address                  object 
 5   city                     object 
 6   zip_code                 float64
 7   store_location           object 
 8   county_number            float64
 9   county                   object 
 10  category                 float64
 11  category_name            object 
 12  vendor_number            float64
 13  vendor_name              object 
 14  item_number              int64  
 15  item_description         object 
 16  pack                     int64  
 17  bottle_volume_ml         int64  
 18  state_bottle_cost        float64
 19  state_bottle_retail      float64
 20  bottles_sold             int64  
 21  sale_dol

#### The original data has 2805307 rows and 24 columns

In [4]:
df.shape

(2805307, 24)

In [5]:
df.nunique()

invoice_and_item_number    2805307
date                           326
store_number                  1954
store_name                    1952
address                       1936
city                           442
zip_code                       479
store_location                2570
county_number                   99
county                         119
category                        57
category_name                   57
vendor_number                  222
vendor_name                    227
item_number                   4545
item_description              4032
pack                            21
bottle_volume_ml                25
state_bottle_cost             1312
state_bottle_retail           1315
bottles_sold                   397
sale_dollars                 11136
volume_sold_liters             772
volume_sold_gallons            761
dtype: int64

###

### Assessment

In [6]:
df.describe()

store_number      zip_code  county_number      category  vendor_number  \
count  2.805307e+06  2.805165e+06   2.805165e+06  2.805307e+06   2.805303e+06   
mean   4.137796e+03  5.124374e+04   5.712714e+01  1.055993e+06   2.871833e+02   
std    1.265248e+03  9.906550e+02   2.732875e+01  1.039322e+05   1.415585e+02   
min    2.106000e+03  5.000200e+04   1.000000e+00  1.011000e+06   3.300000e+01   
25%    2.644000e+03  5.031400e+04   3.100000e+01  1.012100e+06   2.050000e+02   
50%    4.186000e+03  5.104000e+04   6.200000e+01  1.031200e+06   2.600000e+02   
75%    5.244000e+03  5.230200e+04   7.700000e+01  1.062500e+06   4.200000e+02   
max    9.049000e+03  5.722200e+04   9.900000e+01  1.901200e+06   9.780000e+02   

        item_number          pack  bottle_volume_ml  state_bottle_cost  \
count  2.805307e+06  2.805307e+06      2.805307e+06       2.805307e+06   
mean   5.456197e+04  1.194263e+01      8.218059e+02       1.132500e+01   
std    9.185349e+04  7.846285e+00      5.254814e+02       1.103231e+01   
min    2.580000e+02  1.000000e+00      2.000000e+01       6.600000e-01   
25%    2.712500e+04  6.000000e+00      3.750000e+02       6.000000e+00   
50%    3.991600e+04  1.200000e+01      7.500000e+02       8.990000e+00   
75%    6.525100e+04  1.200000e+01      1.000000e+03       1.400000e+01   
max    9.999950e+05  1.200000e+02      5.250000e+03       2.098940e+03   

       state_bottle_retail  bottles_sold  sale_dollars  volume_sold_liters  \
count         2.805307e+06  2.805307e+06  2.805307e+06        2.805307e+06   
mean          1.698823e+01  1.185837e+01  1.624497e+02        9.383131e+00   
std           1.654628e+01  3.566817e+01  5.871846e+02        4.120727e+01   
min           9.900000e-01  1.000000e+00  1.340000e+00        2.000000e-02   
25%           9.000000e+00  3.000000e+00  4.200000e+01        1.500000e+00   
50%           1.349000e+01  6.000000e+00  8.892000e+01        4.500000e+00   
75%           2.100000e+01  1.200000e+01  1.665000e+02        1.050000e+01   
max           3.148410e+03  1.320000e+04  2.509320e+05        1.320000e+04   

       volume_sold_gallons  
count         2.805307e+06  
mean          2.473141e+00  
std           1.088601e+01  
min           0.000000e+00  
25%           3.900000e-01  
50%           1.180000e+00  
75%           2.770000e+00  
max           3.487070e+03

In [7]:
#Because we have e for all outputs, we want to change it
pd.set_option('display.float_format', lambda x: '{:.6f}'.format(x))

df.describe()

store_number       zip_code  county_number       category  \
count 2805307.000000 2805165.000000 2805165.000000 2805307.000000   
mean     4137.795966   51243.743009      57.127137 1055993.194862   
std      1265.247896     990.654998      27.328748  103932.156424   
min      2106.000000   50002.000000       1.000000 1011000.000000   
25%      2644.000000   50314.000000      31.000000 1012100.000000   
50%      4186.000000   51040.000000      62.000000 1031200.000000   
75%      5244.000000   52302.000000      77.000000 1062500.000000   
max      9049.000000   57222.000000      99.000000 1901200.000000   

       vendor_number    item_number           pack  bottle_volume_ml  \
count 2805303.000000 2805307.000000 2805307.000000    2805307.000000   
mean      287.183324   54561.974709      11.942627        821.805932   
std       141.558502   91853.494064       7.846285        525.481377   
min        33.000000     258.000000       1.000000         20.000000   
25%       205.000000   27125.000000       6.000000        375.000000   
50%       260.000000   39916.000000      12.000000        750.000000   
75%       420.000000   65251.000000      12.000000       1000.000000   
max       978.000000  999995.000000     120.000000       5250.000000   

       state_bottle_cost  state_bottle_retail   bottles_sold   sale_dollars  \
count     2805307.000000       2805307.000000 2805307.000000 2805307.000000   
mean           11.324997            16.988232      11.858369     162.449687   
std            11.032310            16.546277      35.668169     587.184640   
min             0.660000             0.990000       1.000000       1.340000   
25%             6.000000             9.000000       3.000000      42.000000   
50%             8.990000            13.490000       6.000000      88.920000   
75%            14.000000            21.000000      12.000000     166.500000   
max          2098.940000          3148.410000   13200.000000  250932.000000   

       volume_sold_liters  volume_sold_gallons  
count      2805307.000000       2805307.000000  
mean             9.383131             2.473141  
std             41.207272            10.886014  
min              0.020000             0.000000  
25%              1.500000             0.390000  
50%              4.500000             1.180000  
75%             10.500000             2.770000  
max          13200.000000          3487.070000

#### Key statistics:
The average number of bottles per order rounds up to 12, with 75% of orders also comprising 12 bottles, indicating consistency. However, the maximum of 13200 bottles suggests a potential outlier. Similarly, the average sales amount, standing at 162.45 US dollars, and in the 75% where sales hover around 166.5 US dollars, reaffirming the consistency observed in most purchase amounts. Yet, a noteworthy anomaly surfaces in the shape of a substantial maximum sales figure, reaching an impressive $250932, which may become another outlier.

###

### Handling Missing Values

#### Now we go to the data cleaning process, let us check null values first.

In [8]:
df.isnull().sum()                                                                                

invoice_and_item_number         0
date                            0
store_number                    0
store_name                      0
address                       142
city                          142
zip_code                      142
store_location             330335
county_number                 142
county                        142
category                        0
category_name                   0
vendor_number                   4
vendor_name                     4
item_number                     0
item_description                0
pack                            0
bottle_volume_ml                0
state_bottle_cost               0
state_bottle_retail             0
bottles_sold                    0
sale_dollars                    0
volume_sold_liters              0
volume_sold_gallons             0
dtype: int64

In [9]:
#check % of missing values
df.isnull().sum()/len(df)*100

invoice_and_item_number    0.000000
date                       0.000000
store_number               0.000000
store_name                 0.000000
address                    0.005062
city                       0.005062
zip_code                   0.005062
store_location            11.775360
county_number              0.005062
county                     0.005062
category                   0.000000
category_name              0.000000
vendor_number              0.000143
vendor_name                0.000143
item_number                0.000000
item_description           0.000000
pack                       0.000000
bottle_volume_ml           0.000000
state_bottle_cost          0.000000
state_bottle_retail        0.000000
bottles_sold               0.000000
sale_dollars               0.000000
volume_sold_liters         0.000000
volume_sold_gallons        0.000000
dtype: float64

###

Given that we possess the address, zip code, county name, and city name in the dataset, including the store name column is unnecessary for our mapping visualizations. Moreover, considering the dataset pertains to Iowa, the map's geographical scope is naturally limited. Therefore, we have decided to exclude the store name from our analysis, as it provides enough information for our mapping needs.

In [10]:
#make a copy just in case, then drop the column
df1= df.drop(columns='store_location')
#sanity check
df1.head()

invoice_and_item_number        date  store_number  \
0         INV-33179700135  2021-01-04          2576   
1         INV-33196200106  2021-01-04          2649   
2         INV-33184300011  2021-01-04          2539   
3         INV-33184100015  2021-01-04          4024   
4         INV-33174200025  2021-01-04          5385   

                             store_name         address             city  \
0  Hy-Vee Wine and Spirits / Storm Lake  1250 N Lake St       Storm Lake   
1                   Hy-Vee #3 / Dubuque   400 Locust St          Dubuque   
2        Hy-Vee Food Store / Iowa Falls      640 S. Oak       Iowa Falls   
3            Wal-Mart 1546 / Iowa Falls       840 S Oak       Iowa Falls   
4                    Vine Food & Liquor   2704 Vine St.  West Des Moines   

      zip_code  county_number      county       category  ... item_number  \
0 50588.000000      11.000000  BUENA VIST 1081600.000000  ...       64870   
1 52001.000000      31.000000     DUBUQUE 1081200.000000  ...       65200   
2 50126.000000      42.000000      HARDIN 1031100.000000  ...       38008   
3 50126.000000      42.000000      HARDIN 1031100.000000  ...       36648   
4 50265.000000      77.000000        POLK 1012200.000000  ...        4626   

            item_description pack  bottle_volume_ml state_bottle_cost  \
0  Fireball Cinnamon Whiskey   48               100          0.900000   
1       Tequila Rose Liqueur   12               750         11.500000   
2         Smirnoff 80prf PET    6              1750         14.750000   
3              Caliber Vodka   12               750          3.310000   
4       Buchanan Deluxe 12YR   12               750         20.990000   

   state_bottle_retail  bottles_sold  sale_dollars  volume_sold_liters  \
0             1.350000            48     64.800000            4.800000   
1            17.250000             4     69.000000            3.000000   
2            22.130000             6    132.780000           10.500000   
3             4.970000            12     59.640000            9.000000   
4            31.490000             2     62.980000            1.500000   

   volume_sold_gallons  
0             1.260000  
1             0.790000  
2             2.770000  
3             2.370000  
4             0.390000  

[5 rows x 23 columns]

Check null values again

In [11]:
rows_with_null = df1[df1.isnull().any(axis=1)]
rows_with_null.head(10)

invoice_and_item_number        date  store_number  \
101913         INV-41995800001  2021-11-15          6229   
102454         INV-41995800027  2021-11-15          6229   
102802         INV-41995800003  2021-11-15          6229   
103286         INV-41995800011  2021-11-15          6229   
103602         INV-41995800014  2021-11-15          6229   
103719         INV-41995800006  2021-11-15          6229   
103735         INV-41995800022  2021-11-15          6229   
103837         INV-41995800017  2021-11-15          6229   
104024         INV-41995800024  2021-11-15          6229   
104268         INV-41995800026  2021-11-15          6229   

                                 store_name address city  zip_code  \
101913  Beer Thirty Storm Lake / Storm Lake     NaN  NaN       NaN   
102454  Beer Thirty Storm Lake / Storm Lake     NaN  NaN       NaN   
102802  Beer Thirty Storm Lake / Storm Lake     NaN  NaN       NaN   
103286  Beer Thirty Storm Lake / Storm Lake     NaN  NaN       NaN   
103602  Beer Thirty Storm Lake / Storm Lake     NaN  NaN       NaN   
103719  Beer Thirty Storm Lake / Storm Lake     NaN  NaN       NaN   
103735  Beer Thirty Storm Lake / Storm Lake     NaN  NaN       NaN   
103837  Beer Thirty Storm Lake / Storm Lake     NaN  NaN       NaN   
104024  Beer Thirty Storm Lake / Storm Lake     NaN  NaN       NaN   
104268  Beer Thirty Storm Lake / Storm Lake     NaN  NaN       NaN   

        county_number county       category  ... item_number  \
101913            NaN    NaN 1701100.000000  ...       21228   
102454            NaN    NaN 1031100.000000  ...       38178   
102802            NaN    NaN 1701100.000000  ...       16906   
103286            NaN    NaN 1062100.000000  ...       42426   
103602            NaN    NaN 1022100.000000  ...       89198   
103719            NaN    NaN 1011200.000000  ...      920374   
103735            NaN    NaN 1082000.000000  ...       65258   
103837            NaN    NaN 1012200.000000  ...        5349   
104024            NaN    NaN 1071100.000000  ...       63753   
104268            NaN    NaN 1081600.000000  ...      100413   

                              item_description pack  bottle_volume_ml  \
101913                         1792 Full Proof    6               750   
102454                    Titos Handmade Vodka    6              1750   
102802                         Bookers Bourbon    6               750   
103286                           Barcelo Anejo   12               750   
103602           Jose Cuervo Especial Reposado    6              1750   
103719                  SOOH Old Forester 1910    6               750   
103735                    Jagermeister Liqueur    6              1750   
103837                      Johnnie Walker Red    6              1750   
104024               Salvadors Peach Margarita    6              1750   
104268  Fireball Cinnamon Whiskey Party Bucket    1                50   

       state_bottle_cost  state_bottle_retail  bottles_sold  sale_dollars  \
101913         24.000000            36.000000             1     36.000000   
102454         19.000000            28.500000           360  10260.000000   
102802         42.500000            63.750000             1     63.750000   
103286          7.450000            11.180000            12    134.160000   
103602         21.500000            32.250000             6    193.500000   
103719         27.990000            41.990000             6    251.940000   
103735         26.050000            39.080000             2     78.160000   
103837         26.500000            39.750000             6    238.500000   
104024          6.000000             9.000000             6     54.000000   
104268         51.600000            77.400000             1     77.400000   

        volume_sold_liters  volume_sold_gallons  
101913            0.750000             0.190000  
102454          630.000000           166.420000  
102802            0.750000             0.190000  
103286            

###

We have chosen to utilize the store name as the key to fill in the null values for address, city, zip code, county number, and county. This decision is grounded in our belief that these attributes are all intricately linked to the unique store names, allowing us to maintain clean data.

###

Check how many unique store names has null values

In [12]:
unique_store_names_with_null = rows_with_null['store_name'].unique()
unique_store_names_with_null

array(['Beer Thirty Storm Lake / Storm Lake', 'Northside Liquor',
       'Benz Distributing', 'Hy-Vee #3 / BDI / Des Moines',
       'Hy-Vee Food Store / Mount Ayr', 'Bootlegging Barzinis',
       "Casey's General Store #2811 / Springville"], dtype=object)

###

In [13]:
#Before fill in null values, make a copy of df1 just in case
df2 = df1.copy()

Check if the name of store has a unique store number

In [14]:
beer_liquor_rows = df1[df1['store_name'] == 'Beer Thirty Storm Lake / Storm Lake'].groupby('store_number')['store_number'].unique()
beer_liquor_rows

store_number
6229    [6229]
Name: store_number, dtype: object

In [15]:
#check what information need to be filled
beer_liquor_rows_with_noNull = df1[df1['store_name'] == 'Beer Thirty Storm Lake / Storm Lake'].dropna(subset=['address'])
beer_liquor_rows_with_noNull

invoice_and_item_number        date  store_number  \
155678          INV-42238900003  2021-11-22          6229   
156362          INV-42238900031  2021-11-22          6229   
156538          INV-42238900001  2021-11-22          6229   
156710          INV-42238900013  2021-11-22          6229   
157071          INV-42238900026  2021-11-22          6229   
...                         ...         ...           ...   
2664153         INV-41317000005  2021-10-25          6229   
2664213         INV-41316800024  2021-10-25          6229   
2664358         INV-41316800030  2021-10-25          6229   
2664438         INV-41316800010  2021-10-25          6229   
2664637         INV-41316900036  2021-10-25          6229   

                                  store_name           address        city  \
155678   Beer Thirty Storm Lake / Storm Lake   208 E Milwaukee  Storm Lake   
156362   Beer Thirty Storm Lake / Storm Lake   208 E Milwaukee  Storm Lake   
156538   Beer Thirty Storm Lake / Storm Lake   208 E Milwaukee  Storm Lake   
156710   Beer Thirty Storm Lake / Storm Lake   208 E Milwaukee  Storm Lake   
157071   Beer Thirty Storm Lake / Storm Lake   208 E Milwaukee  Storm Lake   
...                                      ...               ...         ...   
2664153  Beer Thirty Storm Lake / Storm Lake  205 E 7th Street  Storm Lake   
2664213  Beer Thirty Storm Lake / Storm Lake  205 E 7th Street  Storm Lake   
2664358  Beer Thirty Storm Lake / Storm Lake  205 E 7th Street  Storm Lake   
2664438  Beer Thirty Storm Lake / Storm Lake  205 E 7th Street  Storm Lake   
2664637  Beer Thirty Storm Lake / Storm Lake  205 E 7th Street  Storm Lake   

            zip_code  county_number      county       category  ...  \
155678  50588.000000      11.000000  BUENA VIST 1011200.000000  ...   
156362  50588.000000      11.000000  BUENA VIST 1081200.000000  ...   
156538  50588.000000      11.000000  BUENA VIST 1011300.000000  ...   
156710  50588.000000      11.000000  BUENA VIST 1022200.000000  ...   
157071  50588.000000      11.000000  BUENA VIST 1012100.000000  ...   
...              ...            ...         ...            ...  ...   
2664153 50588.000000      11.000000  BUENA VIST 1031100.000000  ...   
2664213 50588.000000      11.000000  BUENA VIST 1082000.000000  ...   
2664358 50588.000000      11.000000  BUENA VIST 1011600.000000  ...   
2664438 50588.000000      11.000000  BUENA VIST 1022200.000000  ...   
2664637 50588.000000      11.000000  BUENA VIST 1082000.000000  ...   

        item_number                  item_description pack  bottle_volume_ml  \
155678       920376            SOOH Old Forester 1920    6               750   
156362        68036      Baileys Original Irish Cream   12               750   
156538        21589        EH Taylor Jr Single Barrel    6               750   
156710        89496       Margaritaville Gold Tequila   12               750   
157071        11348  Seagrams VO Canadian Whiskey PET    6              1750   
...             ...                               ...  ...               ...   
2664153       36903             McCormick 80prf Vodka   48               200   
2664213       64876                  Drambuie Liqueur   12               750   
2664358       27125                 Templeton Rye 6YR    6               750   
2664438       89841                    Hornitos Anejo   12               750   
2664637       65251         Jagermeister Liqueur Mini    5                50   

        state_bottle_cost  state_bottle_retail  bottles_sold  sale_dollars  \
155678          30.980000            46.470000             6    278.820000   
156362          16.490000            24.740000            12    296.880000   
156538          33.340000            50.010000             1     50.010000   
156710           7.510000            11.270000            12    108.000000   
157071          10.450000            15.680000             6     94.080000   
...                   ...                  ...          

We have identified instances where the same store number is associated with two different addresses. To maintain data integrity, we have decided to use the new address information to fill in the data, ensuring consistency and cleanliness throughout the dataset​.

In [16]:
fill_values = {
    'address': '205 E 7th Street',
    'city': 'Storm Lake',
    'zip_code': 50588.0,
    'county_number': 11.0,
    'county': 'BUENA VIST'
}
# Iterate over rows and fill null values
for index, row in df2.iterrows():
    if row['store_name'] == 'Beer Thirty Storm Lake / Storm Lake':
        for column in fill_values.keys():
            if pd.isnull(row[column]):
                df2.at[index, column] = fill_values[column]

In [17]:
#Sanity Check
beer_liquor_rows = df2[df2['store_name'] == 'Beer Thirty Storm Lake / Storm Lake']['store_name'].isnull().any()
beer_liquor_rows

False

###

Repeat those steps for Northside Liquor

In [18]:
#check how many unique stores does northside liquor store have
northside_liquor_rows = df1[df1['store_name'] == 'Northside Liquor'].groupby('store_number')['store_number'].unique()
northside_liquor_rows

store_number
5251    [5251]
Name: store_number, dtype: object

In [19]:
#check what information need to be filled
northside_liquor_rows = df1[df1['store_name'] == 'Northside Liquor']
northside_liquor_rows

invoice_and_item_number        date  store_number        store_name  \
31725           INV-33285700023  2021-01-07          5251  Northside Liquor   
31738           INV-33285700065  2021-01-07          5251  Northside Liquor   
32049           INV-33285800001  2021-01-07          5251  Northside Liquor   
32106           INV-33285700044  2021-01-07          5251  Northside Liquor   
32300           INV-33285900002  2021-01-07          5251  Northside Liquor   
...                         ...         ...           ...               ...   
1807161         INV-34951900031  2021-03-11          5251  Northside Liquor   
1807187         INV-34951900038  2021-03-11          5251  Northside Liquor   
1807530         INV-34951900004  2021-03-11          5251  Northside Liquor   
1807545         INV-34951900026  2021-03-11          5251  Northside Liquor   
1807578         INV-34951900042  2021-03-11          5251  Northside Liquor   

                    address        city     zip_code  county_number  \
31725    1303 North Federal  Mason City 50401.000000      17.000000   
31738    1303 North Federal  Mason City 50401.000000      17.000000   
32049    1303 North Federal  Mason City 50401.000000      17.000000   
32106    1303 North Federal  Mason City 50401.000000      17.000000   
32300    1303 North Federal  Mason City 50401.000000      17.000000   
...                     ...         ...          ...            ...   
1807161  1303 North Federal  Mason City 50401.000000      17.000000   
1807187  1303 North Federal  Mason City 50401.000000      17.000000   
1807530  1303 North Federal  Mason City 50401.000000      17.000000   
1807545  1303 North Federal  Mason City 50401.000000      17.000000   
1807578  1303 North Federal  Mason City 50401.000000      17.000000   

             county       category  ... item_number  \
31725    CERRO GORD 1022100.000000  ...       89197   
31738    CERRO GORD 1012400.000000  ...       15641   
32049    CERRO GORD 1091300.000000  ...       76414   
32106    CERRO GORD 1071100.000000  ...       63345   
32300    CERRO GORD 1012200.000000  ...        5318   
...             ...            ...  ...         ...   
1807161  CERRO GORD 1071100.000000  ...       59159   
1807187  CERRO GORD 1012100.000000  ...       11774   
1807530  CERRO GORD 1701100.000000  ...       21946   
1807545  CERRO GORD 1031100.000000  ...       37348   
1807578  CERRO GORD 1082000.000000  ...       65253   

                                          item_description pack  \
31725                        Jose Cuervo Especial Reposado   12   
31738                                Jameson Caskmates IPA   24   
32049                          Ole Smoky Moonshine Pickles    6   
32106    On the Rocks Cocktails Tres Gen Jalapeno Pinea...   12   
32300                          Johnnie Walker Double Black    6   
...                                                    ...  ...   
1807161                  1800 Ultimate Raspberry Margarita    6   
1807187                                       Black Velvet   24   
1807530                        Wild Turkey Kentucky Spirit    6   
1807545                                     Phillips Vodka    6   
1807578                               Jagermeister Liqueur   48   

         bottle_volume_ml state_bottle_cost  state_bottle_retail  \
31725                1000         13.000000            19.500000   
31738                 375          6.490000             9.740000   
32049                 750         12.500000            18.750000   
32106                 375          6.000000             9.000000   
32300                 750         24.050000            36.080000   
...                   ...               ...                  ...   
1807161              1750         10.040000            15.060000   
1807187               375          3.070000             4.610000   
1807530               750         30.000000            45.000000   
1807545              1750          7.600000            11.40

In [20]:
# Define the values to fill nulls
fill_values = {
    'address': '1303 North Federal',
    'city': 'Mason City',
    'zip_code': 50401.0,
    'county_number': 17.0,
    'county': 'CERRO GORD'
}

# Iterate over rows and fill null values where 'store_name' is 'Northside Liquor'
for index, row in df2.iterrows():
    if row['store_name'] == 'Northside Liquor':
        for column in fill_values.keys():
            if pd.isnull(row[column]):
                df2.at[index, column] = fill_values[column]

In [21]:
#Sanity Check
North_null_values = df2[df2['store_name'] == 'Northside Liquor']['store_name'].isnull().any()
North_null_values

False

###

Repeat those steps for Benz Distributing

In [22]:
Benz_liquor_rows = df2[df2['store_name'] == 'Benz Distributing'].groupby('store_number')['store_number'].unique()
Benz_liquor_rows

store_number
3773    [3773]
Name: store_number, dtype: object

In [23]:
Benz_liquor_rows = df2[df2['store_name'] == 'Benz Distributing']
Benz_liquor_rows

invoice_and_item_number        date  store_number         store_name  \
239             INV-33170600014  2021-01-04          3773  Benz Distributing   
378             INV-33170700043  2021-01-04          3773  Benz Distributing   
414             INV-33170600043  2021-01-04          3773  Benz Distributing   
425             INV-33170700031  2021-01-04          3773  Benz Distributing   
507             INV-33170600021  2021-01-04          3773  Benz Distributing   
...                         ...         ...           ...                ...   
2784551         INV-41849100033  2021-11-10          3773  Benz Distributing   
2784686         INV-41849100004  2021-11-10          3773  Benz Distributing   
2784836         INV-41849200053  2021-11-10          3773  Benz Distributing   
2784844         INV-41849200011  2021-11-10          3773  Benz Distributing   
2784907         INV-41849100007  2021-11-10          3773  Benz Distributing   

                address          city     zip_code  county_number county  \
239      501 7th Ave SE  Cedar Rapids 52401.000000      57.000000   LINN   
378      501 7th Ave SE  Cedar Rapids 52401.000000      57.000000   LINN   
414      501 7th Ave SE  Cedar Rapids 52401.000000      57.000000   LINN   
425      501 7th Ave SE  Cedar Rapids 52401.000000      57.000000   LINN   
507      501 7th Ave SE  Cedar Rapids 52401.000000      57.000000   LINN   
...                 ...           ...          ...            ...    ...   
2784551  501 7th Ave SE  Cedar Rapids 52401.000000      57.000000   LINN   
2784686  501 7th Ave SE  Cedar Rapids 52401.000000      57.000000   LINN   
2784836  501 7th Ave SE  Cedar Rapids 52401.000000      57.000000   LINN   
2784844  501 7th Ave SE  Cedar Rapids 52401.000000      57.000000   LINN   
2784907  501 7th Ave SE  Cedar Rapids 52401.000000      57.000000   LINN   

              category  ... item_number  \
239     1041100.000000  ...       29515   
378     1081300.000000  ...       75210   
414     1011200.000000  ...       27369   
425     1081100.000000  ...       67527   
507     1011100.000000  ...       16026   
...                ...  ...         ...   
2784551 1091300.000000  ...       76413   
2784686 1701100.000000  ...       17917   
2784836 1082000.000000  ...       64645   
2784844 1062200.000000  ...       43127   
2784907 1032100.000000  ...       34458   

                               item_description pack  bottle_volume_ml  \
239                         Bowling & Burch Gin    6               750   
378                                  Kinky Pink    6               750   
414      Cedar Ridge Port Cask Finished Bourbon    6               750   
425                               Kahlua Coffee   12              1000   
507             Hatozaki Finest Japanese Whisky    6               750   
...                                         ...  ...               ...   
2784551        Ole Smoky Butter Pecan Moonshine    6               750   
2784686              Elijah Craig  Barrel Proof    6               750   
2784836                       Domaine de Canton    6               750   
2784844                        Bacardi Superior   12              1000   
2784907                               Ketel One    6              1750   

        state_bottle_cost  state_bottle_retail  bottles_sold  sale_dollars  \
239             20.000000            30.000000             2     60.000000   
378             10.000000            15.000000            30    450.000000   
414             25.500000            38.250000             2     76.500000   
425             15.990000            23.990000            12    287.880000   
507             23.150000            34.730000             2     69.460000   
...                   ...                  ...           ...           ...   
2784551         12.500000            18.750000             3     56.250000   
2784686         35.000000            52.500000             6    315.000000   
2784836         17.500000 

In [24]:
fill_values = {
    'address': '501 7th Ave SE',
    'city': 'Cedar Rapids',
    'zip_code': 52401.0,
    'county_number': 57.0,
    'county': 'LINN'
}

for index, row in df2.iterrows():
    if row['store_name'] == 'Benz Distributing':
        for column in fill_values.keys():
            if pd.isnull(row[column]):
                df2.at[index, column] = fill_values[column]

In [25]:
Benz_null_values = df2[df2['store_name'] == 'Benz Distributing']['store_name'].isnull().any()
Benz_null_values

False

###

Repeat those steps for Hy-Vee #3 / BDI / Des Moines

In [26]:
Hy3_liquor_rows = df2[df2['store_name'] == 'Hy-Vee #3 / BDI / Des Moines'].groupby('store_number')['store_number'].unique()
Hy3_liquor_rows

store_number
2633    [2633]
Name: store_number, dtype: object

In [27]:
Hy3_liquor_rows = df2[df2['store_name'] == 'Hy-Vee #3 / BDI / Des Moines']
Hy3_liquor_rows

invoice_and_item_number        date  store_number  \
87              INV-33169200066  2021-01-04          2633   
96              INV-33169200117  2021-01-04          2633   
380             INV-33169200051  2021-01-04          2633   
503             INV-33169200091  2021-01-04          2633   
586             INV-33169200061  2021-01-04          2633   
...                         ...         ...           ...   
2793582         INV-41877300005  2021-11-11          2633   
2793617         INV-41878700045  2021-11-11          2633   
2793675         INV-41878700069  2021-11-11          2633   
2793698         INV-41878700063  2021-11-11          2633   
2793726         INV-41878700111  2021-11-11          2633   

                           store_name          address        city  \
87       Hy-Vee #3 / BDI / Des Moines  3221 SE 14th St  Des Moines   
96       Hy-Vee #3 / BDI / Des Moines  3221 SE 14th St  Des Moines   
380      Hy-Vee #3 / BDI / Des Moines  3221 SE 14th St  Des Moines   
503      Hy-Vee #3 / BDI / Des Moines  3221 SE 14th St  Des Moines   
586      Hy-Vee #3 / BDI / Des Moines  3221 SE 14th St  Des Moines   
...                               ...              ...         ...   
2793582  Hy-Vee #3 / BDI / Des Moines  3221 SE 14th St  Des Moines   
2793617  Hy-Vee #3 / BDI / Des Moines  3221 SE 14th St  Des Moines   
2793675  Hy-Vee #3 / BDI / Des Moines  3221 SE 14th St  Des Moines   
2793698  Hy-Vee #3 / BDI / Des Moines  3221 SE 14th St  Des Moines   
2793726  Hy-Vee #3 / BDI / Des Moines  3221 SE 14th St  Des Moines   

            zip_code  county_number county       category  ... item_number  \
87      50320.000000      77.000000   POLK 1081200.000000  ...       68049   
96      50320.000000      77.000000   POLK 1062200.000000  ...       43127   
380     50320.000000      77.000000   POLK 1062500.000000  ...       43050   
503     50320.000000      77.000000   POLK 1081400.000000  ...       82627   
586     50320.000000      77.000000   POLK 1032200.000000  ...       64914   
...              ...            ...    ...            ...  ...         ...   
2793582 50320.000000      77.000000   POLK 1071100.000000  ...       59037   
2793617 50320.000000      77.000000   POLK 1032200.000000  ...       34787   
2793675 50320.000000      77.000000   POLK 1062500.000000  ...       44419   
2793698 50320.000000      77.000000   POLK 1062400.000000  ...       43338   
2793726 50320.000000      77.000000   POLK 1081300.000000  ...       76487   

                                    item_description pack  bottle_volume_ml  \
87                          Baileys Vanilla Cinnamon   12               750   
96                                  Bacardi Superior   12              1000   
380                             Bacardi Dragon Berry   12              1000   
503                           Dekuyper Cherry Pucker   12              1000   
586                                      Ciroc Mango   12               750   
...                                              ...  ...               ...   
2793582  Desert Island Long Island Iced Tea Cocktail   12              1000   
2793617                         Stolichnaya Blueberi   12              1000   
2793675                          Cruzan Black Cherry   12               750   
2793698               Captain Morgan Original Spiced    6              1750   
2793726                           Paramount Amaretto   12              1000   

        state_bottle_cost  state_bottle_retail  bottles_sold  sale_dollars  \
87              16.490000            24.740000            12    296.880000   
96               9.500000            14.250000            36    513.000000   
380              9.500000            14.250000            24    342.000000   
503              7.870000            11.810000            12    141.720000   
586             16.490000            24.740000            12    296.880000   
...                   ...                  ...           ...           ...   
27

In [28]:
fill_values = {
    'address': '3221 SE 14th St',
    'city': 'Des Moines',
    'zip_code': 50320.0,
    'county_number': 77.0,
    'county': 'POLK'
}

for index, row in df2.iterrows():
    if row['store_name'] == 'Hy-Vee #3 / BDI / Des Moines':
        for column in fill_values.keys():
            if pd.isnull(row[column]):
                df2.at[index, column] = fill_values[column]

In [29]:
Hy3_null_values = df2[df2['store_name'] == 'Hy-Vee #3 / BDI / Des Moines']['store_name'].isnull().any()
Hy3_null_values

False

###

Repeat those steps for Hy-Vee Food Store / Mount Ayr

In [30]:
HyF_liquor_rows = df2[df2['store_name'] == 'Hy-Vee Food Store / Mount Ayr'].groupby('store_number')['store_number'].unique()
HyF_liquor_rows

store_number
2658    [2658]
Name: store_number, dtype: object

In [31]:
HyF_liquor_rows = df2[df2['store_name'] == 'Hy-Vee Food Store / Mount Ayr']
HyF_liquor_rows

invoice_and_item_number        date  store_number  \
38711           INV-33301900010  2021-01-08          2658   
38965           INV-33301900033  2021-01-08          2658   
39186           INV-33301900022  2021-01-08          2658   
39550           INV-33301900020  2021-01-08          2658   
39623           INV-33301900004  2021-01-08          2658   
...                         ...         ...           ...   
2804424         INV-41946200013  2021-11-12          2658   
2804467         INV-41946200008  2021-11-12          2658   
2804554         INV-41946200022  2021-11-12          2658   
2804558         INV-41946200012  2021-11-12          2658   
2804985         INV-41946200021  2021-11-12          2658   

                            store_name            address       city  \
38711    Hy-Vee Food Store / Mount Ayr     402 S Hayes St  Mount Ayr   
38965    Hy-Vee Food Store / Mount Ayr     402 S Hayes St  Mount Ayr   
39186    Hy-Vee Food Store / Mount Ayr     402 S Hayes St  Mount Ayr   
39550    Hy-Vee Food Store / Mount Ayr     402 S Hayes St  Mount Ayr   
39623    Hy-Vee Food Store / Mount Ayr     402 S Hayes St  Mount Ayr   
...                                ...                ...        ...   
2804424  Hy-Vee Food Store / Mount Ayr  201 N Fillmore St  Mount Ayr   
2804467  Hy-Vee Food Store / Mount Ayr  201 N Fillmore St  Mount Ayr   
2804554  Hy-Vee Food Store / Mount Ayr  201 N Fillmore St  Mount Ayr   
2804558  Hy-Vee Food Store / Mount Ayr  201 N Fillmore St  Mount Ayr   
2804985  Hy-Vee Food Store / Mount Ayr  201 N Fillmore St  Mount Ayr   

            zip_code  county_number    county       category  ... item_number  \
38711   50854.000000      80.000000  RINGGOLD 1062400.000000  ...       43244   
38965   50854.000000      80.000000  RINGGOLD 1012100.000000  ...       11588   
39186   50854.000000      80.000000  RINGGOLD 1081600.000000  ...       64865   
39550   50854.000000      80.000000  RINGGOLD 1011400.000000  ...       74801   
39623   50854.000000      80.000000  RINGGOLD 1031100.000000  ...       36904   
...              ...            ...       ...            ...  ...         ...   
2804424 50854.000000      80.000000  RINGGOLD 1011200.000000  ...       17206   
2804467 50854.000000      80.000000  RINGGOLD 1011100.000000  ...       87026   
2804554 50854.000000      80.000000  RINGGOLD 1701100.000000  ...      100636   
2804558 50854.000000      80.000000  RINGGOLD 1011200.000000  ...       19066   
2804985 50854.000000      80.000000  RINGGOLD 1062500.000000  ...       44499   

                         item_description pack  bottle_volume_ml  \
38711    Captain Morgan 100prf Spiced Rum   12               750   
38965                Black Velvet Reserve    6              1750   
39186       Fireball Cinnamon Whiskey PET   12               750   
39550        Jack Daniels Tennessee Apple   12               750   
39623           McCormick 80prf Vodka PET   24               375   
...                                   ...  ...               ...   
2804424               Cedar Ridge Bourbon    6               750   
2804467   Skrewball Peanut Butter Whiskey    6               750   
2804554          Jim Beam Mini 12 Seasons    6                50   
2804558                          Jim Beam   12               750   
2804985                      Cruzan Mango   12               750   

        state_bottle_cost  state_bottle_retail  bottles_sold  sale_dollars  \
38711           10.490000            15.740000            12    188.880000   
38965           12.990000            19.490000             6    116.940000   
39186            8.980000            13.470000            12    161.640000   
39550           15.570000            23.360000             2     46.720000   
39623            1.800000             2.700000            24     64.800000   
...                   ...                  ...           ...           ...   
2804424         18.100000            27.150000             6    162.900000

In [32]:
fill_values = {
    'address': '201 N Fillmore St',
    'city': 'Mount Ayr',
    'zip_code': 50854.0,
    'county_number': 80.0,
    'county': 'RINGGOLD'
}

for index, row in df2.iterrows():
    if row['store_name'] == 'Hy-Vee Food Store / Mount Ayr':
        for column in fill_values.keys():
            if pd.isnull(row[column]):
                df2.at[index, column] = fill_values[column]

###

Repeat those steps for Bootlegging Barzinis

In [33]:
Boot_liquor_rows = df2[df2['store_name'] == 'Bootlegging Barzinis'].groupby('store_number')['store_number'].unique()
Boot_liquor_rows

store_number
6035    [6035]
Name: store_number, dtype: object

In [34]:
Boot_liquor_rows = df2[df2['store_name'] == 'Bootlegging Barzinis']
Boot_liquor_rows

invoice_and_item_number        date  store_number  \
31760           INV-33272300036  2021-01-07          6035   
31835           INV-33272300041  2021-01-07          6035   
31843           INV-33272300029  2021-01-07          6035   
31905           INV-33272100006  2021-01-07          6035   
31908           INV-33272200001  2021-01-07          6035   
...                         ...         ...           ...   
2774379         INV-41815900006  2021-11-09          6035   
2774464         INV-41816100075  2021-11-09          6035   
2774686         INV-41816100027  2021-11-09          6035   
2775382         INV-41816100003  2021-11-09          6035   
2775406         INV-41816100051  2021-11-09          6035   

                   store_name        address        city     zip_code  \
31760    Bootlegging Barzinis  412 First Ave  Coralville 52241.000000   
31835    Bootlegging Barzinis  412 First Ave  Coralville 52241.000000   
31843    Bootlegging Barzinis  412 First Ave  Coralville 52241.000000   
31905    Bootlegging Barzinis  412 First Ave  Coralville 52241.000000   
31908    Bootlegging Barzinis  412 First Ave  Coralville 52241.000000   
...                       ...            ...         ...          ...   
2774379  Bootlegging Barzinis  412 First Ave  Coralville 52241.000000   
2774464  Bootlegging Barzinis  412 First Ave  Coralville 52241.000000   
2774686  Bootlegging Barzinis  412 First Ave  Coralville 52241.000000   
2775382  Bootlegging Barzinis  412 First Ave  Coralville 52241.000000   
2775406  Bootlegging Barzinis  412 First Ave  Coralville 52241.000000   

         county_number   county       category  ... item_number  \
31760        52.000000  JOHNSON 1082000.000000  ...       65144   
31835        52.000000  JOHNSON 1011400.000000  ...       87043   
31843        52.000000  JOHNSON 1082000.000000  ...       64876   
31905        52.000000  JOHNSON 1012300.000000  ...      994972   
31908        52.000000  JOHNSON 1082000.000000  ...       64776   
...                ...      ...            ...  ...         ...   
2774379      52.000000  JOHNSON 1082100.000000  ...      966067   
2774464      52.000000  JOHNSON 1011600.000000  ...      101296   
2774686      52.000000  JOHNSON 1031200.000000  ...       41989   
2775382      52.000000  JOHNSON 1701100.000000  ...       25459   
2775406      52.000000  JOHNSON 1022100.000000  ...       89234   

                                          item_description pack  \
31760                               Il Tramonto Limoncello    6   
31835                      Ole Smoky Peanut Butter Whiskey    6   
31843                                     Drambuie Liqueur   12   
31905                            Glendronach 12YR Original    6   
31908                                    Cointreau Liqueur   12   
...                                                    ...  ...   
2774379                                     Toschi Nocello    6   
2774464                       WhistlePig 6YR PiggyBack Rye    4   
2774686                                            UV Cake   12   
2775382  Slipknot Limited Edition Anniversary Iowa Labe...    6   
2775406                                      Corazon Anejo    6   

         bottle_volume_ml state_bottle_cost  state_bottle_retail  \
31760                 750         10.500000            15.750000   
31835                 750         10.000000            15.000000   
31843                 750         23.000000            34.500000   
31905                 750         32.500000            48.750000   
31908                 750         19.990000            29.990000   
...                   ...               ...                  ...   
2774379               750         17.670000            26.510000   
2774464                50         49.410000            74.120000   
2774686               750          7.000000            10.500000   
2775382               750         25.000000            37.500000   
2775406               750         15.000

In [35]:
fill_values = {
    'address': '412 First Ave',
    'city': 'Coralville',
    'zip_code': 52241.0,
    'county_number': 52.0,
    'county': 'JOHNSON'
}

for index, row in df2.iterrows():
    if row['store_name'] == 'Bootlegging Barzinis  ':
        for column in fill_values.keys():
            if pd.isnull(row[column]):
                df2.at[index, column] = fill_values[column]

###

Repeat those steps for Casey's General Store #2811 / Springville

In [36]:
Casey_liquor_rows = df2[df2['store_name'] == "Casey's General Store #2811 / Springville"].groupby('store_number')['store_number'].unique()
Casey_liquor_rows

store_number
4934    [4934]
Name: store_number, dtype: object

In [37]:
Casey_liquor_rows = df2[df2['store_name'] == "Casey's General Store #2811 / Springville"]
Casey_liquor_rows

invoice_and_item_number        date  store_number  \
236815          INV-37221200007  2021-06-05          4934   
237326          INV-37221200003  2021-06-05          4934   
238233          INV-37221300005  2021-06-05          4934   
239212          INV-37221300003  2021-06-05          4934   
239762          INV-37221300014  2021-06-05          4934   
...                         ...         ...           ...   
2803344         INV-41952100015  2021-11-12          4934   
2803590         INV-41952100006  2021-11-12          4934   
2804277         INV-41952100003  2021-11-12          4934   
2804573         INV-41952100012  2021-11-12          4934   
2804635         INV-41952100011  2021-11-12          4934   

                                        store_name       address         city  \
236815   Casey's General Store #2811 / Springville  605 6th St S  Springville   
237326   Casey's General Store #2811 / Springville  605 6th St S  Springville   
238233   Casey's General Store #2811 / Springville  605 6th St S  Springville   
239212   Casey's General Store #2811 / Springville  605 6th St S  Springville   
239762   Casey's General Store #2811 / Springville  605 6th St S  Springville   
...                                            ...           ...          ...   
2803344  Casey's General Store #2811 / Springville  605 6th St S  Springville   
2803590  Casey's General Store #2811 / Springville  605 6th St S  Springville   
2804277  Casey's General Store #2811 / Springville  605 6th St S  Springville   
2804573  Casey's General Store #2811 / Springville  605 6th St S  Springville   
2804635  Casey's General Store #2811 / Springville  605 6th St S  Springville   

            zip_code  county_number county       category  ... item_number  \
236815  52336.000000      57.000000   LINN 1081600.000000  ...       64865   
237326  52336.000000      57.000000   LINN 1081300.000000  ...       75230   
238233  52336.000000      57.000000   LINN 1062400.000000  ...       43338   
239212  52336.000000      57.000000   LINN 1062200.000000  ...       43125   
239762  52336.000000      57.000000   LINN 1062400.000000  ...       43024   
...              ...            ...    ...            ...  ...         ...   
2803344 52336.000000      57.000000   LINN 1081600.000000  ...       65013   
2803590 52336.000000      57.000000   LINN 1011100.000000  ...       25616   
2804277 52336.000000      57.000000   LINN 1012100.000000  ...       10805   
2804573 52336.000000      57.000000   LINN 1081600.000000  ...       64866   
2804635 52336.000000      57.000000   LINN 1031200.000000  ...       39492   

                              item_description pack  bottle_volume_ml  \
236815           Fireball Cinnamon Whiskey PET   12               750   
237326                               Kinky Red    6               750   
238233          Captain Morgan Original Spiced    6              1750   
239212                    Bacardi Superior PET   12               750   
239762                   Admiral Nelson Spiced   24               375   
...                                        ...  ...               ...   
2803344  Fireball Cinnamon Whiskey Mini Sleeve   12                50   
2803590             Seagrams 7 Crown PET Flask   12               750   
2804277                Crown Royal Regal Apple   24               375   
2804573              Fireball Cinnamon Whiskey   12               750   
2804635             New Amsterdam Pink Whitney   12               750   

        state_bottle_cost  state_bottle_retail  bottles_sold  sale_dollars  \
236815           8.980000            13.470000            12    161.640000   
237326          10.000000            15.000000             6     90.000000   
238233          18.000000            27.000000             6    162.000000   
239212           8.260000            12.390000            12    148.680000   
239762           2.990000             4.490000            24    107.760000   
...                 

In [38]:
fill_values = {
    'address': '605 6th St S',
    'city': 'Springville',
    'zip_code': 52336.0,
    'county_number': 57.0,
    'county': 'LINN'
}

for index, row in df2.iterrows():
    if row['store_name'] == "Casey's General Store #2811 / Springville":
        for column in fill_values.keys():
            if pd.isnull(row[column]):
                df2.at[index, column] = fill_values[column]

###

After we run all the loops to fill in null values, we check those columns

In [39]:
df2.isnull().sum()

invoice_and_item_number    0
date                       0
store_number               0
store_name                 0
address                    0
city                       0
zip_code                   0
county_number              0
county                     0
category                   0
category_name              0
vendor_number              4
vendor_name                4
item_number                0
item_description           0
pack                       0
bottle_volume_ml           0
state_bottle_cost          0
state_bottle_retail        0
bottles_sold               0
sale_dollars               0
volume_sold_liters         0
volume_sold_gallons        0
dtype: int64

Now we only have 8 null values in total.  We will keep trying to fill them in.

###

In [40]:
#Run this code to see all columns
pd.set_option('display.max_columns', None)

In [41]:
df2.head()

invoice_and_item_number        date  store_number  \
0         INV-33179700135  2021-01-04          2576   
1         INV-33196200106  2021-01-04          2649   
2         INV-33184300011  2021-01-04          2539   
3         INV-33184100015  2021-01-04          4024   
4         INV-33174200025  2021-01-04          5385   

                             store_name         address             city  \
0  Hy-Vee Wine and Spirits / Storm Lake  1250 N Lake St       Storm Lake   
1                   Hy-Vee #3 / Dubuque   400 Locust St          Dubuque   
2        Hy-Vee Food Store / Iowa Falls      640 S. Oak       Iowa Falls   
3            Wal-Mart 1546 / Iowa Falls       840 S Oak       Iowa Falls   
4                    Vine Food & Liquor   2704 Vine St.  West Des Moines   

      zip_code  county_number      county       category    category_name  \
0 50588.000000      11.000000  BUENA VIST 1081600.000000  Whiskey Liqueur   
1 52001.000000      31.000000     DUBUQUE 1081200.000000   Cream Liqueurs   
2 50126.000000      42.000000      HARDIN 1031100.000000  American Vodkas   
3 50126.000000      42.000000      HARDIN 1031100.000000  American Vodkas   
4 50265.000000      77.000000        POLK 1012200.000000  Scotch Whiskies   

   vendor_number               vendor_name  item_number  \
0     421.000000      SAZERAC COMPANY  INC        64870   
1     300.000000  McCormick Distilling Co.        65200   
2     260.000000           DIAGEO AMERICAS        38008   
3      55.000000     SAZERAC NORTH AMERICA        36648   
4     260.000000           DIAGEO AMERICAS         4626   

            item_description  pack  bottle_volume_ml  state_bottle_cost  \
0  Fireball Cinnamon Whiskey    48               100           0.900000   
1       Tequila Rose Liqueur    12               750          11.500000   
2         Smirnoff 80prf PET     6              1750          14.750000   
3              Caliber Vodka    12               750           3.310000   
4       Buchanan Deluxe 12YR    12               750          20.990000   

   state_bottle_retail  bottles_sold  sale_dollars  volume_sold_liters  \
0             1.350000            48     64.800000            4.800000   
1            17.250000             4     69.000000            3.000000   
2            22.130000             6    132.780000           10.500000   
3             4.970000            12     59.640000            9.000000   
4            31.490000             2     62.980000            1.500000   

   volume_sold_gallons  
0             1.260000  
1             0.790000  
2             2.770000  
3             2.370000  
4             0.390000

In [42]:
#Get the null rows
null_vendor_rows = df2[df2['vendor_name'].isnull() | df2['vendor_number'].isnull()]
pd.set_option('display.max_columns', None)
null_vendor_rows

invoice_and_item_number        date  store_number  \
337624          INV-35139200005  2021-03-18          5251   
459143          INV-33740700003  2021-01-25          3773   
465554          INV-33728600005  2021-01-25          2633   
1397930         INV-35973800003  2021-04-20          6035   

                           store_name             address          city  \
337624               Northside Liquor  1303 North Federal    Mason City   
459143              Benz Distributing      501 7th Ave SE  Cedar Rapids   
465554   Hy-Vee #3 / BDI / Des Moines     3221 SE 14th St    Des Moines   
1397930          Bootlegging Barzinis       412 First Ave    Coralville   

            zip_code  county_number      county       category  \
337624  50401.000000      17.000000  CERRO GORD 1032100.000000   
459143  52401.000000      57.000000        LINN 1012400.000000   
465554  50320.000000      77.000000        POLK 1022200.000000   
1397930 52241.000000      52.000000     JOHNSON 1012400.000000   

              category_name  vendor_number vendor_name  item_number  \
337624      Imported Vodkas            NaN         NaN       965108   
459143       Irish Whiskies            NaN         NaN       915574   
465554   100% Agave Tequila            NaN         NaN       988100   
1397930      Irish Whiskies            NaN         NaN       915574   

           item_description  pack  bottle_volume_ml  state_bottle_cost  \
337624        Grey Goose VX     6              1000          31.340000   
459143    Connemara 12 Year     6               750          14.660000   
465554   Monte Alban Silver    12               750          12.980000   
1397930      Connemara 12YR     6               750          30.840000   

         state_bottle_retail  bottles_sold  sale_dollars  volume_sold_liters  \
337624             80.000000            18   1440.000000           18.000000   
459143             56.250000             6    337.500000            4.500000   
465554             15.560000            12    186.720000            9.000000   
1397930            56.250000             6    337.500000            4.500000   

         volume_sold_gallons  
337624              4.750000  
459143              1.180000  
465554              2.370000  
1397930             1.180000

###

In [43]:
#Check if we can use invoice numbers to get vendor names and fill them in
invoice_numbers = ['INV-35139200005', 'INV-33740700003', 'INV-33728600005', 'INV-35973800003']
filtered_rows = df2[df2['invoice_and_item_number'].isin(invoice_numbers)]
filtered_rows

invoice_and_item_number        date  store_number  \
337624          INV-35139200005  2021-03-18          5251   
459143          INV-33740700003  2021-01-25          3773   
465554          INV-33728600005  2021-01-25          2633   
1397930         INV-35973800003  2021-04-20          6035   

                           store_name             address          city  \
337624               Northside Liquor  1303 North Federal    Mason City   
459143              Benz Distributing      501 7th Ave SE  Cedar Rapids   
465554   Hy-Vee #3 / BDI / Des Moines     3221 SE 14th St    Des Moines   
1397930          Bootlegging Barzinis       412 First Ave    Coralville   

            zip_code  county_number      county       category  \
337624  50401.000000      17.000000  CERRO GORD 1032100.000000   
459143  52401.000000      57.000000        LINN 1012400.000000   
465554  50320.000000      77.000000        POLK 1022200.000000   
1397930 52241.000000      52.000000     JOHNSON 1012400.000000   

              category_name  vendor_number vendor_name  item_number  \
337624      Imported Vodkas            NaN         NaN       965108   
459143       Irish Whiskies            NaN         NaN       915574   
465554   100% Agave Tequila            NaN         NaN       988100   
1397930      Irish Whiskies            NaN         NaN       915574   

           item_description  pack  bottle_volume_ml  state_bottle_cost  \
337624        Grey Goose VX     6              1000          31.340000   
459143    Connemara 12 Year     6               750          14.660000   
465554   Monte Alban Silver    12               750          12.980000   
1397930      Connemara 12YR     6               750          30.840000   

         state_bottle_retail  bottles_sold  sale_dollars  volume_sold_liters  \
337624             80.000000            18   1440.000000           18.000000   
459143             56.250000             6    337.500000            4.500000   
465554             15.560000            12    186.720000            9.000000   
1397930            56.250000             6    337.500000            4.500000   

         volume_sold_gallons  
337624              4.750000  
459143              1.180000  
465554              2.370000  
1397930             1.180000

Since invoice numbers are unique and cannot be used to obtain vendor information, and considering we have only 4 rows with null values, we have opted to drop these rows to ensure data accuracy and reliability.

###

In [44]:
#make a new df just in case
df3 = df2.drop(filtered_rows.index)
df3.head()

invoice_and_item_number        date  store_number  \
0         INV-33179700135  2021-01-04          2576   
1         INV-33196200106  2021-01-04          2649   
2         INV-33184300011  2021-01-04          2539   
3         INV-33184100015  2021-01-04          4024   
4         INV-33174200025  2021-01-04          5385   

                             store_name         address             city  \
0  Hy-Vee Wine and Spirits / Storm Lake  1250 N Lake St       Storm Lake   
1                   Hy-Vee #3 / Dubuque   400 Locust St          Dubuque   
2        Hy-Vee Food Store / Iowa Falls      640 S. Oak       Iowa Falls   
3            Wal-Mart 1546 / Iowa Falls       840 S Oak       Iowa Falls   
4                    Vine Food & Liquor   2704 Vine St.  West Des Moines   

      zip_code  county_number      county       category    category_name  \
0 50588.000000      11.000000  BUENA VIST 1081600.000000  Whiskey Liqueur   
1 52001.000000      31.000000     DUBUQUE 1081200.000000   Cream Liqueurs   
2 50126.000000      42.000000      HARDIN 1031100.000000  American Vodkas   
3 50126.000000      42.000000      HARDIN 1031100.000000  American Vodkas   
4 50265.000000      77.000000        POLK 1012200.000000  Scotch Whiskies   

   vendor_number               vendor_name  item_number  \
0     421.000000      SAZERAC COMPANY  INC        64870   
1     300.000000  McCormick Distilling Co.        65200   
2     260.000000           DIAGEO AMERICAS        38008   
3      55.000000     SAZERAC NORTH AMERICA        36648   
4     260.000000           DIAGEO AMERICAS         4626   

            item_description  pack  bottle_volume_ml  state_bottle_cost  \
0  Fireball Cinnamon Whiskey    48               100           0.900000   
1       Tequila Rose Liqueur    12               750          11.500000   
2         Smirnoff 80prf PET     6              1750          14.750000   
3              Caliber Vodka    12               750           3.310000   
4       Buchanan Deluxe 12YR    12               750          20.990000   

   state_bottle_retail  bottles_sold  sale_dollars  volume_sold_liters  \
0             1.350000            48     64.800000            4.800000   
1            17.250000             4     69.000000            3.000000   
2            22.130000             6    132.780000           10.500000   
3             4.970000            12     59.640000            9.000000   
4            31.490000             2     62.980000            1.500000   

   volume_sold_gallons  
0             1.260000  
1             0.790000  
2             2.770000  
3             2.370000  
4             0.390000

In [45]:
#sanity check
df3.isna().sum()

invoice_and_item_number    0
date                       0
store_number               0
store_name                 0
address                    0
city                       0
zip_code                   0
county_number              0
county                     0
category                   0
category_name              0
vendor_number              0
vendor_name                0
item_number                0
item_description           0
pack                       0
bottle_volume_ml           0
state_bottle_cost          0
state_bottle_retail        0
bottles_sold               0
sale_dollars               0
volume_sold_liters         0
volume_sold_gallons        0
dtype: int64

###

### Extra-Cleaning

In [47]:
#check all county names
county_names = df3['county'].unique()

# Convert the array to a list and print
print(list(county_names))

['BUENA VIST', 'DUBUQUE', 'HARDIN', 'POLK', 'BUTLER', 'BENTON', 'MUSCATINE', 'LINN', 'CLINTON', 'SAC', 'MITCHELL', 'POWESHIEK', 'SCOTT', 'JOHNSON', 'CEDAR', 'BLACK HAWK', 'FRANKLIN', 'Linn', 'CARROLL', 'OBRIEN', 'GRUNDY', 'KOSSUTH', 'STORY', 'CHICKASAW', 'CLARKE', 'MAHASKA', 'IOWA', 'JACKSON', 'IDA', 'WARREN', 'LEE', 'CHEROKEE', 'GREENE', 'WORTH', 'HOWARD', 'WEBSTER', 'CALHOUN', 'CERRO GORD', 'Polk', 'MARSHALL', 'FLOYD', 'WAPELLO', 'HAMILTON', 'WOODBURY', 'MARION', 'TAMA', 'LYON', 'Delaware', 'DALLAS', 'FAYETTE', 'JASPER', 'SHELBY', 'PLYMOUTH', 'AUDUBON', 'MONONA', 'CRAWFORD', 'BUCHANAN', 'SIOUX', 'WINNESHIEK', 'DELAWARE', 'HARRISON', 'Black Hawk', 'BREMER', 'CLAYTON', 'Webster', 'OSCEOLA', 'Dallas', 'JONES', 'JEFFERSON', 'PALO ALTO', 'POTTAWATTA', 'KEOKUK', 'HANCOCK', 'WRIGHT', 'DICKINSON', 'CLAY', 'WASHINGTON', 'EMMET', 'WINNEBAGO', 'VAN BUREN', 'Wapello', 'MILLS', 'DAVIS', 'HENRY', 'Hancock', 'WAYNE', 'GUTHRIE', 'APPANOOSE', 'ADAMS', 'ADAIR', 'MONROE', 'DES MOINES', 'ALLAMAKEE', 'Ma

In [48]:
#print outputs alphabetically
sorted_county_names = sorted(list(county_names))
print(sorted_county_names)

['ADAIR', 'ADAMS', 'ALLAMAKEE', 'APPANOOSE', 'AUDUBON', 'Adair', 'BENTON', 'BLACK HAWK', 'BOONE', 'BREMER', 'BUCHANAN', 'BUENA VIST', 'BUTLER', 'Black Hawk', 'CALHOUN', 'CARROLL', 'CASS', 'CEDAR', 'CERRO GORD', 'CHEROKEE', 'CHICKASAW', 'CLARKE', 'CLAY', 'CLAYTON', 'CLINTON', 'CRAWFORD', 'DALLAS', 'DAVIS', 'DECATUR', 'DELAWARE', 'DES MOINES', 'DICKINSON', 'DUBUQUE', 'Dallas', 'Delaware', 'Des Moines', 'EMMET', 'FAYETTE', 'FLOYD', 'FRANKLIN', 'FREMONT', 'GREENE', 'GRUNDY', 'GUTHRIE', 'HAMILTON', 'HANCOCK', 'HARDIN', 'HARRISON', 'HENRY', 'HOWARD', 'HUMBOLDT', 'Hancock', 'Hardin', 'Henry', 'IDA', 'IOWA', 'Iowa', 'JACKSON', 'JASPER', 'JEFFERSON', 'JOHNSON', 'JONES', 'Jackson', 'KEOKUK', 'KOSSUTH', 'LEE', 'LINN', 'LOUISA', 'LUCAS', 'LYON', 'Linn', 'MADISON', 'MAHASKA', 'MARION', 'MARSHALL', 'MILLS', 'MITCHELL', 'MONONA', 'MONROE', 'MONTGOMERY', 'MUSCATINE', 'Madison', 'Marion', 'Marshall', 'OBRIEN', 'OSCEOLA', 'PAGE', 'PALO ALTO', 'PLYMOUTH', 'POCAHONTAS', 'POLK', 'POTTAWATTA', 'POWESHIEK', 

###

We found that the name of counties are in different format.  We need to fix it.

In [56]:
#make a copy just in case
df4 = df3.copy()
df4.head(10)

invoice_and_item_number        date  store_number  \
0         INV-33179700135  2021-01-04          2576   
1         INV-33196200106  2021-01-04          2649   
2         INV-33184300011  2021-01-04          2539   
3         INV-33184100015  2021-01-04          4024   
4         INV-33174200025  2021-01-04          5385   
5         INV-33186700007  2021-01-04          4110   
6         INV-33197500003  2021-01-04          4228   
7         INV-33197200010  2021-01-04          2713   
8         INV-33174600126  2021-01-04          2648   
9         INV-33202000002  2021-01-04          5735   

                             store_name              address             city  \
0  Hy-Vee Wine and Spirits / Storm Lake       1250 N Lake St       Storm Lake   
1                   Hy-Vee #3 / Dubuque        400 Locust St          Dubuque   
2        Hy-Vee Food Store / Iowa Falls           640 S. Oak       Iowa Falls   
3            Wal-Mart 1546 / Iowa Falls            840 S Oak       Iowa Falls   
4                    Vine Food & Liquor        2704 Vine St.  West Des Moines   
5                 Brothers Market, Inc.       706 Highway 57      Parkersburg   
6          Fareway Stores #462 / Vinton            501 A Ave           Vinton   
7        Hy-Vee Dyersville Dollar Fresh     1201 12th Ave SE       Dyersville   
8                       Hy-Vee #4 / WDM        555 S 51st St  West Des Moines   
9         Super Saver Liquor -Muscatine  1510 A Isett Avenue        Muscatine   

      zip_code  county_number      county       category  \
0 50588.000000      11.000000  BUENA VIST 1081600.000000   
1 52001.000000      31.000000     DUBUQUE 1081200.000000   
2 50126.000000      42.000000      HARDIN 1031100.000000   
3 50126.000000      42.000000      HARDIN 1031100.000000   
4 50265.000000      77.000000        POLK 1012200.000000   
5 50665.000000      12.000000      BUTLER 1032100.000000   
6 52349.000000       6.000000      BENTON 1032100.000000   
7 52040.000000      31.000000     DUBUQUE 1031100.000000   
8 50265.000000      77.000000        POLK 1012200.000000   
9 52761.000000      70.000000   MUSCATINE 1091300.000000   

                    category_name  vendor_number               vendor_name  \
0                 Whiskey Liqueur     421.000000      SAZERAC COMPANY  INC   
1                  Cream Liqueurs     300.000000  McCormick Distilling Co.   
2                 American Vodkas     260.000000           DIAGEO AMERICAS   
3                 American Vodkas      55.000000     SAZERAC NORTH AMERICA   
4                 Scotch Whiskies     260.000000           DIAGEO AMERICAS   
5                 Imported Vodkas     115.000000  CONSTELLATION BRANDS INC   
6                 Imported Vodkas     370.000000         PERNOD RICARD USA   
7                 American Vodkas     434.000000                 LUXCO INC   
8                 Scotch Whiskies     260.000000           DIAGEO AMERICAS   
9  Neutral Grain Spirits Flavored     346.000000  OLE SMOKY DISTILLERY LLC   

   item_number                          item_description  pack  \
0        64870                 Fireball Cinnamon Whiskey    48   
1        65200                      Tequila Rose Liqueur    12   
2        38008                        Smirnoff 80prf PET     6   
3        36648                             Caliber Vodka    12   
4         4626                      Buchanan Deluxe 12YR    12   
5        34821                              Svedka 80prf     6   
6        34006               Absolut Swedish Vodka 80prf    12   
7        36308                             Hawkeye Vodka     6   
8         5318               Johnnie Walker Double Black     6   
9        86739  Ole Smoky Apple Pie Moonshine 70prf Mini     8   

   bottle_volume_ml  state_bottle_cost  state_bottle_retail  bottles_sold  \
0               100           0.900000             1.350000            48   
1               750          11.500000            17.250000             4   
2              1750       

In [49]:
# Capitalize and sort county names
capitalized_counties = [county.capitalize() for county in county_names]
sorted_counties = sorted(capitalized_counties)

# Loop through sorted counties and print
for county in sorted_counties:
    print(county)

Adair
Adair
Adams
Allamakee
Appanoose
Audubon
Benton
Black hawk
Black hawk
Boone
Bremer
Buchanan
Buena vist
Butler
Calhoun
Carroll
Cass
Cedar
Cerro gord
Cherokee
Chickasaw
Clarke
Clay
Clayton
Clinton
Crawford
Dallas
Dallas
Davis
Decatur
Delaware
Delaware
Des moines
Des moines
Dickinson
Dubuque
Emmet
Fayette
Floyd
Franklin
Fremont
Greene
Grundy
Guthrie
Hamilton
Hancock
Hancock
Hardin
Hardin
Harrison
Henry
Henry
Howard
Humboldt
Ida
Iowa
Iowa
Jackson
Jackson
Jasper
Jefferson
Johnson
Jones
Keokuk
Kossuth
Lee
Linn
Linn
Louisa
Lucas
Lyon
Madison
Madison
Mahaska
Marion
Marion
Marshall
Marshall
Mills
Mitchell
Monona
Monroe
Montgomery
Muscatine
Obrien
Osceola
Page
Palo alto
Plymouth
Pocahontas
Polk
Polk
Pottawatta
Pottawatta
Poweshiek
Poweshiek
Ringgold
Sac
Scott
Scott
Shelby
Sioux
Story
Tama
Taylor
Union
Van buren
Wapello
Wapello
Warren
Washington
Wayne
Webster
Webster
Winnebago
Winneshiek
Woodbury
Worth
Wright


In [51]:
# Create a mapping dictionary to map original county names to capitalized and sorted names
county_mapping = dict(zip(county_names, sorted_counties))
county_mapping

{'BUENA VIST': 'Adair',
 'DUBUQUE': 'Adair',
 'HARDIN': 'Adams',
 'POLK': 'Allamakee',
 'BUTLER': 'Appanoose',
 'BENTON': 'Audubon',
 'MUSCATINE': 'Benton',
 'LINN': 'Black hawk',
 'CLINTON': 'Black hawk',
 'SAC': 'Boone',
 'MITCHELL': 'Bremer',
 'POWESHIEK': 'Buchanan',
 'SCOTT': 'Buena vist',
 'JOHNSON': 'Butler',
 'CEDAR': 'Calhoun',
 'BLACK HAWK': 'Carroll',
 'FRANKLIN': 'Cass',
 'Linn': 'Cedar',
 'CARROLL': 'Cerro gord',
 'OBRIEN': 'Cherokee',
 'GRUNDY': 'Chickasaw',
 'KOSSUTH': 'Clarke',
 'STORY': 'Clay',
 'CHICKASAW': 'Clayton',
 'CLARKE': 'Clinton',
 'MAHASKA': 'Crawford',
 'IOWA': 'Dallas',
 'JACKSON': 'Dallas',
 'IDA': 'Davis',
 'WARREN': 'Decatur',
 'LEE': 'Delaware',
 'CHEROKEE': 'Delaware',
 'GREENE': 'Des moines',
 'WORTH': 'Des moines',
 'HOWARD': 'Dickinson',
 'WEBSTER': 'Dubuque',
 'CALHOUN': 'Emmet',
 'CERRO GORD': 'Fayette',
 'Polk': 'Floyd',
 'MARSHALL': 'Franklin',
 'FLOYD': 'Fremont',
 'WAPELLO': 'Greene',
 'HAMILTON': 'Grundy',
 'WOODBURY': 'Guthrie',
 'MARION': 

In [60]:
# Map the county names in subset1_df using the county_mapping dictionary
df4['capitalized_county'] = df3['county'].map(county_mapping)

In [62]:
#sanity check
df4.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2805303 entries, 0 to 2805306
Data columns (total 24 columns):
 #   Column                   Dtype  
---  ------                   -----  
 0   invoice_and_item_number  object 
 1   date                     object 
 2   store_number             int64  
 3   store_name               object 
 4   address                  object 
 5   city                     object 
 6   zip_code                 float64
 7   county_number            float64
 8   county                   object 
 9   category                 float64
 10  category_name            object 
 11  vendor_number            float64
 12  vendor_name              object 
 13  item_number              int64  
 14  item_description         object 
 15  pack                     int64  
 16  bottle_volume_ml         int64  
 17  state_bottle_cost        float64
 18  state_bottle_retail      float64
 19  bottles_sold             int64  
 20  sale_dollars             float64
 21  volume_s

###

In [63]:
#Check Duplicates
duplicate_rows = df4.duplicated()
num_duplicate_rows = duplicate_rows.sum()
num_duplicate_rows

0

Now our data is clean.

In [65]:
pip install pyarrow


     --------------------------------------- 24.3/24.3 MB 11.3 MB/s eta 0:00:00


In [66]:
# Save the cleaned data to a new file
import pyarrow.parquet as pq

# Export DataFrame to Parquet file
df4.to_parquet('IowaClean', engine='pyarrow')